browser/auto_import_muon.cc

```diff
@@ -0,0 +1,129 @@
+#include "brave/common/brave_switches.h"
+
+#include "base/command_line.h"
+#include "base/logging.h"
+#include "base/run_loop.h"
+#include "chrome/browser/browser_process.h"
+#include "chrome/browser/importer/external_process_importer_host.h"
+#include "chrome/browser/importer/importer_list.h"
+#include "chrome/browser/importer/importer_progress_observer.h"
+#include "chrome/browser/profiles/profile_manager.h"
+#include "chrome/common/importer/importer_data_types.h"
+
+namespace brave {
+
+// Copied outright from first_run.cc
+class ImportEndedObserver : public importer::ImporterProgressObserver {
+ public:
+  ImportEndedObserver() : ended_(false) {}
+  ~ImportEndedObserver() override {}
+
+  // importer::ImporterProgressObserver:
+  void ImportStarted() override {}
+  void ImportItemStarted(importer::ImportItem item) override {}
+  void ImportItemEnded(importer::ImportItem item) override {}
+  void ImportEnded() override {
+    ended_ = true;
+    if (!callback_for_import_end_.is_null())
+      callback_for_import_end_.Run();
+  }
+
+  void set_callback_for_import_end(const base::Closure& callback) {
+    callback_for_import_end_ = callback;
+  }
+
+  bool ended() const {
+    return ended_;
+  }
+
+ private:
+  // Set if the import has ended.
+  bool ended_;
+
+  base::Closure callback_for_import_end_;
+
+  DISALLOW_COPY_AND_ASSIGN(ImportEndedObserver);
+};
+
+void ImportFromSourceProfile(const importer::SourceProfile& source_profile,
+                             Profile* target_profile,
+                             uint16_t items_to_import) {
+  // Deletes itself
+  ExternalProcessImporterHost* importer_host =
+      new ExternalProcessImporterHost;
+  // Don't show the warning dialog if import fails
+  importer_host->set_headless();
+
+  ImportEndedObserver observer;
+  importer_host->set_observer(&observer);
+  importer_host->StartImportSettings(source_profile,
+                                     target_profile,
+                                     items_to_import,
+                                     new ProfileWriter(target_profile));
+  // If the import process has not errored out, block on it.
+  if (!observer.ended()) {
+    base::RunLoop loop;
+    observer.set_callback_for_import_end(loop.QuitClosure());
+    loop.Run();
+    observer.set_callback_for_import_end(base::Closure());
+  }
+}
+
+void AutoImportMuon() {
+  LOG(INFO) << "In brave::AutoImportMuon";
+  base::CommandLine& command_line =
+      *base::CommandLine::ForCurrentProcess();
+  if (!command_line.HasSwitch(switches::kUpgradeFromMuon))
+    return;
+
+  LOG(INFO) << "Starting auto-import from Muon";
+
+  // ImporterList::DetectSourceProfiles
+  // -> profiles_loaded_callback
+  // based on https://cs.chromium.org/chromium/src/chrome/browser/first_run/first_run.cc?l=560-577&rcl=e2569db721bb3bd6b12991d581c60933bf555f30
+  base::RunLoop run_loop;
+  auto importer_list = std::make_unique<ImporterList>();
+  importer_list->DetectSourceProfiles(
```

Comment:
we need to be careful about this because we only want to import the primary profile from browser-laptop and that actually won't be in the place that you would expect for Chrome. In chrome it would be user_data_dir/Default, but because of some historical reasons with Electron the default (and only) profile we want to import from muon is actually in the root of the user_data_dir

Reply:
@bridiver This is handled in each of the platform-specific implementations of GetBraveUserDataFolder.